# CHAPTER - 6
# MERGING DATA FRAMES WITH PANDAS

## 6.1. PREPARING DATA

## 6.1.1 Import Tools in Pandas

There are many different data import tools in pandas that we can use to import a data and create dataframe.

1. pd.read_csv()

2. pd.read_excel()

3. pd.read_html()

4. pd.read_json()

## 6.1.2 Different techniques to `read multiple files`

To read multiple files into a datafarme we could use:

### 1. Using separte data farmes

 e.g 
     
     dataframe0 = pd.read_csv('sales-Jan-2015.csv')
 
     dataframe1 = pd.read_csv('sales-Feb-2015.csv')

In [23]:
import pandas as pd

In [31]:
df0 = pd.read_csv('sales-Feb-2015.csv')
df1 = pd.read_csv('sales-Jan-2015.csv')

In [33]:
df0.head()

date          type  year  avg_price   size    nb_sold
0  12/27/2015  conventional  2015       0.95  small  9630000.0
1  12/20/2015  conventional  2015       0.98  small  8710000.0
2  12/13/2015       organic  2015       0.93  small  9860000.0
3   12/6/2015  conventional  2015       0.89  small  9410000.0
4  11/29/2015  conventional  2015       0.99  small  8100000.0

In [34]:
df1.head()

date          type  year  avg_price   size    nb_sold
0  12/27/2015  conventional  2015       0.95  small  9630000.0
1  12/20/2015  conventional  2015       0.98  small  8710000.0
2  12/13/2015       organic  2015       0.93  small  9860000.0
3   12/6/2015  conventional  2015       0.89  small  9410000.0
4  11/29/2015  conventional  2015       0.99  small  8100000.0

### 2. Using a loop: Itrate over a list of filenames

First create a list of filenames

Second initialize an empty list DataFrame

Third - I terate over list of filenames

e.g 
     
     filenames = ['sales-Jan-2015.csv','sales-Feb-2015.csv']
     
     dataframes= []
     
     for f in filenames:
         dataframes.append(pd.read_csv(f))
     
#### We can also use list comprehension to make it more efficient:

    e.g 
    
    filenames = ['sales-Jan-2015.csv','sales-Feb-2015.csv']
     
    dataframes= [pd.read_csv(f) for f in filenames]

In [41]:
filenames = ['sales-Feb-2015.csv','sales-Jan-2015.csv']

dfes= []

for f in filenames:
     dfes.append(pd.read_csv(f)) # creates a list of dataframes called defs

### 3. Using a glob built-in module

When file names `have similar pattern` we can use a **glob()** function from a glob module.

e.g. 

    from glob import glob
    filenames = glob('sales*.csv')
    dataframes = [pd.read_csv(f) for f in filenames]
    
The astrik (*) is a wild card* matches zero or more characters

Matches anything with `prefix` sales and `suffix` .csv

glob() function creates an iterable object

In [49]:
from glob import glob
filenames = glob('sales*.csv')
dataframes = [pd.read_csv(f) for f in filenames] # creates a list of dataframes called dataframes

## 6.1.3 Reindexing DataFrames

We can import many files into a single DataFrame! But before we do so, we make sure that each DataFrame is consistent and have the same format or alignment or sorting.

Indexes are means by which data frames are labelled. So this labelling using indexing is very important when it comes to combining DataFrames.

Let's import some weather data.

In [2]:
import pandas as pd

In [3]:
w_mean = pd.read_csv('quarterly_mean_temp.csv', index_col='Month')
w_mean

Max TempF
Month           
Apr    61.956044
Jan    32.133333
Jul    68.934783
Oct    43.434783

In [4]:
w_max = pd.read_csv('quarterly_max_temp.csv', index_col='Month')
w_max

Max TempF
Month           
Jan           68
Apr           89
Jul           91
Oct           84

As we can see above **the index in w_mean** is **ordered by** `alphabetical order` whereas As **the index in w_max** is **ordered by** `chronological order`. 

#### Issues like this must be handled before we combine dataFrames!!!!

In the above case of DataFrames a chronological order makes sense, so we might want to make both dataframes sorted by a `chronological order`.

## Handling order of Indexes in a DataFrames: Ordering Indexes

We must handle the ordering of indexes in the DataFrames before we actually merge DataFrames!

To handle the ordering of indexes we use a `.reindex()` method!

We can achieve ordiering of indexes using two techniques:

**1. Using Ordered values of indexes in a list**

Here we define a list that contains an ordered-list-of-values of an index. 

The `w_mean` dataframe is ordered by alpahbetical order rather than chronological. 

**2. Using another DataFrame as a Template** 

We can use another DataFrame which is sorted the way we wanted to sort other DataFrames. 

In such cases we can use it as a template to sort other indexs too!

### 1. Using Ordered values of indexes in a list

So we can order the index using ordered values of indexes as follows:

1. First define the order of indexes by a list. 

2. Then pass the list to `.reindex()` method of the dataframe.

In [5]:
w_mean

Max TempF
Month           
Apr    61.956044
Jan    32.133333
Jul    68.934783
Oct    43.434783

In [6]:
# define the order of indexes by a list.

order = ['Jan', 'Apr', 'Jul', 'Oct']

In [7]:
# pass the list to .reindex() method of the dataframe

w_mean_ordered = w_mean.reindex(order)
w_mean_ordered

Max TempF
Month           
Jan    32.133333
Apr    61.956044
Jul    68.934783
Oct    43.434783

Nice! As we can see here now, our `w_mean` dataFrame is ordered the way we wanted it to be!

#### Q. What would happend if we define an index value that is absent in our dataFrame to be Ordered??

Let's check this out:

In [70]:
order = ['Jan','Feb','Mar', 'Apr','May','Jun','Jul','Aug','Sep', 'Oct','Nov', 'Dec']
w_mean_ordered_NaN = w_mean.reindex(order)
w_mean_ordered_NaN

Max TempF
Month           
Jan    32.133333
Feb          NaN
Mar          NaN
Apr    61.956044
May          NaN
Jun          NaN
Jul    68.934783
Aug          NaN
Sep          NaN
Oct    43.434783
Nov          NaN
Dec          NaN

As we can see here, the ordering will be done, but for the missing index value in our DataFrame a NaN value will be inserted. Which means Not a Number. 

### Q. How do we fix such issues?? 

Ans: `.dropna()` Mthod

In [72]:
# Fix the issue with missing values

order = ['Jan','Feb','Mar', 'Apr','May','Jun','Jul','Aug','Sep', 'Oct','Nov', 'Dec']
w_mean_ordered_fix = w_mean.reindex(order).dropna()
w_mean_ordered_fix

Max TempF
Month           
Jan    32.133333
Apr    61.956044
Jul    68.934783
Oct    43.434783

### 2. Using another DataFrame as a Template 

If we have `Perfectly-Ordered-DataFrame` whose index is ordered in a way that we want to order other DataFrames, we can use it to order other dataFrames. 

In this technique, we don't have to manually create an ordered list of values of indexes like the above technique.

In our example DataFrames, w_mean was alphabetically ordered and w_max was chronologically ordered. 

So here our goal is to sort chronolgically, so we can use `w_max` DataFrame as a template to order the indexs of `w_mean` DataFrame.

In [73]:
# our template DF
w_max 

Max TempF
Month           
Jan           68
Apr           89
Jul           91
Oct           84

In [75]:
w_mean

Max TempF
Month           
Apr    61.956044
Jan    32.133333
Jul    68.934783
Oct    43.434783

### Q. So how do we do the ordering using a template DF??
Trick: Simply pass the index of Template DataFrame into a DataFrame to be ordered.

Pass the `index attribute` of the `w_max` to `.reindex` method!

In [78]:
w_mean_ordered_by_DF = w_mean.reindex(w_max.index)
w_mean_ordered_by_DF

Max TempF
Month           
Jan    32.133333
Apr    61.956044
Jul    68.934783
Oct    43.434783

In [79]:
w_max

Max TempF
Month           
Jan           68
Apr           89
Jul           91
Oct           84

Well, well, well! Now as we can see both DataFrames have a chronologically ordered Index!

From here, merging of the two DataFrames will be streight forward!!

### ORDER MATTERS! 

Note that:
   
    w_mean_ordered_by_DF = w_mean.reindex(w_max.index)
    
is NOT SAME AS:

    w_max_ordered_by_DF = w_max.reindex(w_mean.index)
    
  
The order determines which DF is a template and which is going to be sorted!!!!!!

In [88]:
w_mean_ordered_by_DF = w_mean.reindex(w_max.index)
w_mean_ordered_by_DF

Max TempF
Month           
Jan    32.133333
Apr    61.956044
Jul    68.934783
Oct    43.434783

In [89]:
w_max_ordered_by_DF = w_max.reindex(w_mean.index)
w_max_ordered_by_DF

Max TempF
Month           
Apr           89
Jan           68
Jul           91
Oct           84

### Q. What if we want to Recover the original DataFrame after it is being sorted???

To recover the original DataFrame we can simply use `.sort_index()` method `on the sorted DataFrame`.

Let's `un-sort` **w_mean_ordered_by_DF**

In [81]:
w_mean_ordered_by_DF.sort_index()

Max TempF
Month           
Apr    61.956044
Jan    32.133333
Jul    68.934783
Oct    43.434783

In [82]:
w_mean

Max TempF
Month           
Apr    61.956044
Jan    32.133333
Jul    68.934783
Oct    43.434783

As we can see, we recovered our original DataFrame!!

## 6.1.4 Arithmetic with Series & DataFrames

In [24]:
import pandas as pd

In [25]:
weather = pd.read_csv('Pittsburg.csv',index_col='Date', parse_dates=True)
weather.head()

Temperature_C  Precipitation_mm  Min TempF  Max TempF  Mean TempF
Date                                                                         
2020-03-24       0.152572               0.3         66         79        72.5
2020-03-25      -0.347428               0.5         66         84        75.0
2020-03-26      -0.597428               0.1         71         86        78.5
2020-03-27      -0.777428               0.2         70         86        78.0
2020-03-28      -0.987428               0.1         69         86        77.5

In [26]:
### Let's load someweather data
#Slicing and then indexing

weather.loc['2020-3-24':'2020-3-29', 'Precipitation_mm']

Date
2020-03-24    0.3
2020-03-25    0.5
2020-03-26    0.1
2020-03-27    0.2
2020-03-28    0.1
2020-03-29    0.5
Name: Precipitation_mm, dtype: float64

### Scalar Multiplication and Broadcasting

Let's change the Precipitation_mm to cm

In [34]:
# subset the 'Precipitation_mm' column

Precipitation_mm = weather[['Precipitation_mm']]
Precipitation_mm

Precipitation_mm
Date                        
2020-03-24               0.3
2020-03-25               0.5
2020-03-26               0.1
2020-03-27               0.2
2020-03-28               0.1
2020-03-29               0.5
2020-03-30               0.2
2020-04-01               0.2
2020-04-02               0.1
2020-04-03               0.3

In [35]:
# sacalr multiplication: convert to cm

Precipitation_cm = Precipitation_mm*100
Precipitation_cm

Precipitation_mm
Date                        
2020-03-24              30.0
2020-03-25              50.0
2020-03-26              10.0
2020-03-27              20.0
2020-03-28              10.0
2020-03-29              50.0
2020-03-30              20.0
2020-04-01              20.0
2020-04-02              10.0
2020-04-03              30.0

In [36]:
#change the column name to cm

Precipitation_cm.columns = Precipitation_cm.columns.str.replace('mm', 'cm')
Precipitation_cm

Precipitation_cm
Date                        
2020-03-24              30.0
2020-03-25              50.0
2020-03-26              10.0
2020-03-27              20.0
2020-03-28              10.0
2020-03-29              50.0
2020-03-30              20.0
2020-04-01              20.0
2020-04-02              10.0
2020-04-03              30.0

As we can see the `scalar multiplication` is applied to each element, Broadcasting!

In [44]:
week1_range = weather.loc['2020-3-24':'2020-3-29',['Min TempF', 'Max TempF']]
week1_range

Min TempF  Max TempF
Date                            
2020-03-24         66         79
2020-03-25         66         84
2020-03-26         71         86
2020-03-27         70         86
2020-03-28         69         86
2020-03-29         70         89

### .divide() Method

Let's calculate the percentage variation of the temperature for the last week of April 2020. i.e from 2020-03-24 to 2020-03-29.

To do so:

Divide both `Min TempF` col and `Max TempF` col with `Mean TempF` col and multiply the result with 100 i.e 

divide `week1_range` DataFrame by `week1_mean` DataFrame and multiply the result by 100!

In [46]:
week1_range #DataFrame

Min TempF  Max TempF
Date                            
2020-03-24         66         79
2020-03-25         66         84
2020-03-26         71         86
2020-03-27         70         86
2020-03-28         69         86
2020-03-29         70         89

In [45]:
week1_mean = weather.loc['2020-3-24':'2020-3-29','Mean TempF']
week1_mean #series

Date
2020-03-24    72.5
2020-03-25    75.0
2020-03-26    78.5
2020-03-27    78.0
2020-03-28    77.5
2020-03-29    79.5
Name: Mean TempF, dtype: float64

In [48]:
week1_range / week1_mean  #Dataframe /Series

2020-03-24 00:00:00  2020-03-25 00:00:00  2020-03-26 00:00:00  \
Date                                                                        
2020-03-24                  NaN                  NaN                  NaN   
2020-03-25                  NaN                  NaN                  NaN   
2020-03-26                  NaN                  NaN                  NaN   
2020-03-27                  NaN                  NaN                  NaN   
2020-03-28                  NaN                  NaN                  NaN   
2020-03-29                  NaN                  NaN                  NaN   

            2020-03-27 00:00:00  2020-03-28 00:00:00  2020-03-29 00:00:00  \
Date                                                                        
2020-03-24                  NaN                  NaN                  NaN   
2020-03-25                  NaN                  NaN                  NaN   
2020-03-26                  NaN                  NaN                  NaN   
2020-03-27                  NaN                  NaN                  NaN   
2020-03-28                  NaN                  NaN                  NaN   
2020-03-29                  NaN                  NaN                  NaN   

            Max TempF  Min TempF  
Date                              
2020-03-24        NaN        NaN  
2020-03-25        NaN        NaN  
2020-03-26        NaN        NaN  
2020-03-27        NaN        NaN  
2020-03-28        NaN        NaN  
2020-03-29        NaN        NaN

As we can see this is not working!

**Q. Why isn't this working??**

The column values don't match!

**Q. What else can we do??**

We have to use `.divide()` method with axis='rows'

In [49]:
week1_range.divide(week1_mean , axis='rows')

Min TempF  Max TempF
Date                            
2020-03-24   0.910345   1.089655
2020-03-25   0.880000   1.120000
2020-03-26   0.904459   1.095541
2020-03-27   0.897436   1.102564
2020-03-28   0.890323   1.109677
2020-03-29   0.880503   1.119497

`.divide()` **broadcasts** each element in the col of week1_mean to week1_range to produce the desired ratio.

We can see that the temprature varies by about 10 percent in that week!

### Arithematic Oprations with NON-ALIGNED INDEXES

In [12]:
import pandas as pd
bronze = pd.read_csv('bronze.csv', index_col=0) 
# or we can also index by using name of the col
# bronze = pd.read_csv('bronze.csv', index_col='country')
bronze

Total
country       
USA       1052
Soviet     584
UK         505
France     475
Germany    454

In [13]:
silver = pd.read_csv('silver.csv', index_col=0) 
silver

Total
country       
USA       1195
Soviet     627
UK         591
France     461
Italy      394

In [14]:
gold = pd.read_csv('gold.csv', index_col=0) 
gold

Total
country       
USA       2088
Soviet     838
UK         498
Italy      460
Germany    407

### Plus oprator-Vs- `.add()` Method

Let's try adding using the regular `+ oprator`

In [15]:
bronze + silver

Total
country        
France    936.0
Germany     NaN
Italy       NaN
Soviet   1211.0
UK       1096.0
USA      2247.0

As we can see here, when two series being added have NON-ALIGNED indexes, or if their indexes don't match NaN will result. 

This is beacuse arithematic oprations between pandas series is carried out `for rows that have the same index values`!

#### Q. So what's a solution for to handle addition of NON-ALIGNED indexes??

We have to use `.add()` method with `fill_value=0` argument!

Note with out `fill_value=0` we will get the same result as above!!

In [17]:
#with out fill_value=0

bronze.add(silver)

Total
country        
France    936.0
Germany     NaN
Italy       NaN
Soviet   1211.0
UK       1096.0
USA      2247.0

In [18]:
# with fill_value=0
bronze.add(silver, fill_value=0)

Total
country        
France    936.0
Germany   454.0
Italy     394.0
Soviet   1211.0
UK       1096.0
USA      2247.0

Very Good! Now even though the index values are NON-ALIGNING, using .add() method the correct sum has been computed!!

### chaining .add() method

In [19]:
# using the + oprator
bronze + silver + gold

Total
country        
France      NaN
Germany     NaN
Italy       NaN
Soviet   2049.0
UK       1594.0
USA      4335.0

In [20]:
#chaining  the .add() method

bronze.add(silver).add(gold)

Total
country        
France      NaN
Germany     NaN
Italy       NaN
Soviet   2049.0
UK       1594.0
USA      4335.0

In [21]:
#chaining  the .add() method and fill_value=0
bronze.add(silver, fill_value=0).add(gold, fill_value=0)

Total
country        
France    936.0
Germany   861.0
Italy     854.0
Soviet   2049.0
UK       1594.0
USA      4335.0

### CONCLUSION

Just as the `.divide()` Method is more flexeable than the `/` oprator for division, the `.add()` method is more flexeable than the `+ oprator` for addition!!

# 6.2. CONCATENATING DATA

6.2.1. Appending & concatenating Series

6.2.2. Appending & concatenating DataFrames

6.2.3. Concatenation, keys, & MultiIndexes

6.2.4. Outer & inner joins

## 6.2.1. Appending & concatenating Series

We can append and concatenate series together by using :

`.append()` method or

`pd.concat()` Function

In [52]:
# Series of US states

import pandas as pd
northeast = pd.Series(['CT', 'ME', 'MA', 'NH', 'RI', 'VT','NJ', 'NY', 'PA'])
south = pd.Series(['DE', 'FL', 'GA', 'MD', 'NC', 'SC', 'VA','DC', 'WV', 'AL', 'KY', 'MS', 'TN', 'AR', 'LA', 'OK', 'TX'])
midwest = pd.Series(['IL', 'IN', 'MN', 'MO', 'NE', 'ND','SD', 'IA', 'KS', 'MI', 'OH', 'WI'])
west = pd.Series(['AZ', 'CO', 'ID', 'MT', 'NV', 'NM','UT', 'WY', 'AK', 'CA', 'HI', 'OR','WA'])

In [48]:
northeast

0    CT
1    ME
2    MA
3    NH
4    RI
5    VT
6    NJ
7    NY
8    PA
dtype: object

### Appending Series with nonunique Indices Using .append() Method

Generic syntax:
    
    s1.append(s2)

Rows of s2 stacked under rows of s1.

This method works both with series and DataFrames!

In [53]:
east = northeast.append(south)
east

0     CT
1     ME
2     MA
3     NH
4     RI
5     VT
6     NJ
7     NY
8     PA
0     DE
1     FL
2     GA
3     MD
4     NC
5     SC
6     VA
7     DC
8     WV
9     AL
10    KY
11    MS
12    TN
13    AR
14    LA
15    OK
16    TX
dtype: object

**Notice:** The append method stacks rows without adjusting the index values!

Accesing a given value using a given index returns all the values from the respective series that has been appended!

In [54]:
east.loc[0]

0    CT
0    DE
dtype: object

### Appending with unique index values: `.reset_index()`

Usually it is important to have `unique index values.`

`.reset_index(drop=True)` 

**drop=True** : discards the repeated INDEX VALUES and replaces with a new INDEX VALUE but the values corrosponding to each index will reamin unchanged even if it is a repeated value!!!!

In [55]:
new_east = northeast.append(south).reset_index(drop=True)
new_east

0     CT
1     ME
2     MA
3     NH
4     RI
5     VT
6     NJ
7     NY
8     PA
9     DE
10    FL
11    GA
12    MD
13    NC
14    SC
15    VA
16    DC
17    WV
18    AL
19    KY
20    MS
21    TN
22    AR
23    LA
24    OK
25    TX
dtype: object

### Using concat() Function

The function concat can do an equialent task just like append.

Generic syntax: 

    pd.concat([s1,s2])
    
The required input is a list of `series` or `DataFrame.`

In [56]:
east_con = pd.concat([northeast,south])
east_con # also equivalent with east

0     CT
1     ME
2     MA
3     NH
4     RI
5     VT
6     NJ
7     NY
8     PA
0     DE
1     FL
2     GA
3     MD
4     NC
5     SC
6     VA
7     DC
8     WV
9     AL
10    KY
11    MS
12    TN
13    AR
14    LA
15    OK
16    TX
dtype: object

### using conact with unique index vales: `ignore_index=True` Argument

Just like we used `.reset_index(drop=True)` on our append method to return unique index values, we can use `ignore_index=True` Argument with concat function to get unique index values.

In [57]:
new_east_con = pd.concat([northeast,south], ignore_index=True)
new_east_con # also equivalent with east

0     CT
1     ME
2     MA
3     NH
4     RI
5     VT
6     NJ
7     NY
8     PA
9     DE
10    FL
11    GA
12    MD
13    NC
14    SC
15    VA
16    DC
17    WV
18    AL
19    KY
20    MS
21    TN
22    AR
23    LA
24    OK
25    TX
dtype: object

Ads we can see we are able to achieve the same effect using concat function as well!

## 6.2.2. Appending & concatenating DataFrames

Let's load population data first.

In [62]:
import pandas as pd
pop1 = pd.read_csv('pop1.csv', index_col=0)
pop1

# row label: Zip code
# col: population in the given zip code area

2010  Sensus Population
Zip Code ZCTA                         
66407                              479
72732                             4716
50579                             2405
46241                            30670

In [60]:
pop2 = pd.read_csv('pop2.csv', index_col=0)
pop2

2010  Sensus Population
Zip Code ZCTA                         
12776                             2180
76092                             2669
98360                             1221
49464                            27481

In [61]:
print(type(pop2), pop2.shape)

<class 'pandas.core.frame.DataFrame'> (4, 1)


### Stacking DataFrames with the same column names and index names:

Let's stack pop1 DataFrame over pop2 DataFrame!

In [64]:
# using append:

p1_app_p2 = pop1.append(pop2)
p1_app_p2

2010  Sensus Population
Zip Code ZCTA                         
66407                              479
72732                             4716
50579                             2405
46241                            30670
12776                             2180
76092                             2669
98360                             1221
49464                            27481

In [66]:
# usning concat func

p1_con_p2 = pd.concat([pop1,pop2])
p1_con_p2

2010  Sensus Population
Zip Code ZCTA                         
66407                              479
72732                             4716
50579                             2405
46241                            30670
12776                             2180
76092                             2669
98360                             1221
49464                            27481

As we can see we get the same result suing `pop1.append(pop2)` and `pd.concat([pop1, pop2])`

### Stacking DataFrames with the DIFFERENT column names and index names:

In [74]:
import pandas as pd

unemployment = pd.read_csv('unemployment.csv', index_col=0)
unemployment

unemployment  participants
Zip                              
2860           0.11         34447
46167          0.02          4800
1097           0.33            42
80808          0.07          4310

In [75]:
population = pd.read_csv('pop.csv', index_col=0)
population

2010  Sensus Population
Zip Code ZCTA                         
57538                              322
59916                              130
37660                            40038
2860                             45199

In [79]:
## Let's stack population over unemployment
## Vertical stacking
pu_app = population.append(unemployment)
pu_app 

2010  Sensus Population  participants  unemployment
57538                    322.0           NaN           NaN
59916                    130.0           NaN           NaN
37660                  40038.0           NaN           NaN
2860                   45199.0           NaN           NaN
2860                       NaN       34447.0          0.11
46167                      NaN        4800.0          0.02
1097                       NaN          42.0          0.33
80808                      NaN        4310.0          0.07

### concatenating rows:  stacking by row

`axis=0` is a default value, if we don't specify it is by default will be zero.

Vertical stacking

In [78]:
#Equivalently we can also use conact: Vertical stacking i.e pop at the top and unemp at bottm

pu_con = pd.concat([population,unemployment], axis=0)  
pu_con 

2010  Sensus Population  participants  unemployment
57538                    322.0           NaN           NaN
59916                    130.0           NaN           NaN
37660                  40038.0           NaN           NaN
2860                   45199.0           NaN           NaN
2860                       NaN       34447.0          0.11
46167                      NaN        4800.0          0.02
1097                       NaN          42.0          0.33
80808                      NaN        4310.0          0.07

The `top 4 rows` are from the `population` DataFrame and 
The `bottom 4-rows` are from the `unemployment` DataFrame.

**NOTICE:** We have repeated index lebels `2860` the first came from `population` DataFrame and the second came from `unemployment` DataFrame.

### concatenating columns:  stacking by column

`axis=1` 

horizontal stacking

In [80]:
pu_con = pd.concat([population,unemployment], axis=1)  
pu_con 

2010  Sensus Population  unemployment  participants
1097                       NaN          0.33          42.0
2860                   45199.0          0.11       34447.0
37660                  40038.0           NaN           NaN
46167                      NaN          0.02        4800.0
57538                    322.0           NaN           NaN
59916                    130.0           NaN           NaN
80808                      NaN          0.07        4310.0

**NOTICE:** We `DO NOT` have repeated index lebels for `2860` i.e. only unique index values are listed!

This is also called `OUTER JOIN`

## 6.2.3. Concatenation, keys, & MultiIndexes

Often we want to concatenate DataFrames that have exat same row labels and column names.

Let's import such data:

In [81]:
import pandas as pd

In [82]:
file1 = 'q1_rainfall_2013.csv'
rain2013 = pd.read_csv(file1, index_col='Month', parse_dates=True)
rain2013

Precipitation
Month               
Jan         0.096129
Feb         0.067143
Mar         0.061613

In [83]:
file2 = 'q1_rainfall_2014.csv'
rain2014 = pd.read_csv(file2, index_col='Month', parse_dates=True)
rain2014

Precipitation
Month               
Jan         0.050323
Feb         0.082143
Mar         0.070968

As we can see rain2013 and rain2014 have exact same **row labels** : `Jan Feb Mar` and **column name** as `Precipitation`.

### Conactenating Rows: Vertical stacking

Let's stack rain2013 over rain2014.

In [85]:
rain1314_by_row =pd.concat([rain2013, rain2014], axis=0)
rain1314_by_row

Precipitation
Month               
Jan         0.096129
Feb         0.067143
Mar         0.061613
Jan         0.050323
Feb         0.082143
Mar         0.070968

As we can see here we have a `repeating row lebels` and **we don't know which data belongs to which year!**

Q. How dow we fix this problem?

Ans: Using multi-index on rows

### conactenating Using multi-index on rows

In [88]:
rain1314_by_row_MIR =pd.concat([rain2013, rain2014],keys=[2013,2014], axis=0)
rain1314_by_row_MIR

Precipitation
     Month               
2013 Jan         0.096129
     Feb         0.067143
     Mar         0.061613
2014 Jan         0.050323
     Feb         0.082143
     Mar         0.070968

Well now it's vivid which data belongs to which year!!

We can easily the multi-index with loc as:

In [89]:
rain1314_by_row_MIR.loc[2013]

Precipitation
Month               
Jan         0.096129
Feb         0.067143
Mar         0.061613

### Conactenating Columns: Horizontal stacking

Let's concatentae rain2014 next to rain2013.

In [90]:
rain1314_by_col =pd.concat([rain2013, rain2014], axis=1)
rain1314_by_col

Precipitation  Precipitation
Month                              
Jan         0.096129       0.050323
Feb         0.067143       0.082143
Mar         0.061613       0.070968

Here also as we can see we have repeating column values, we don't know which Precipitation data belongs to which year!!

Q. How do we fix this??

Ans: conactenating Using multi-index on cols

### conactenating Using multi-index on cols

In [91]:
rain1314_by_row_MIC =pd.concat([rain2013, rain2014],keys=[2013,2014], axis=1)
rain1314_by_row_MIC

2013          2014
      Precipitation Precipitation
Month                            
Jan        0.096129      0.050323
Feb        0.067143      0.082143
Mar        0.061613      0.070968

As we can see here we have a multi level column labeling which makes it more clear that which data belongs to which year.

To access simply pass the column name:

In [92]:
rain1314_by_row_MIC[2013]

Precipitation
Month               
Jan         0.096129
Feb         0.067143
Mar         0.061613

## Conact() with Dictionary

As we have seen so far concat() function accepts a list value.

concat() function also accepts a dictionary value.

Let's implement the multi-index concatenation using dictionaries.

In [93]:
file1 = 'q1_rainfall_2013.csv'
file2 = 'q1_rainfall_2013.csv'
rain2013 = pd.read_csv(file1, index_col='Month', parse_dates=True)
rain2014 = pd.read_csv(file2, index_col='Month', parse_dates=True)

In [95]:
rain_dict = {2013:rain2013 ,  2014:rain2014}
print(rain_dict)

{2013:        Precipitation
Month               
Jan         0.096129
Feb         0.067143
Mar         0.061613, 2014:        Precipitation
Month               
Jan         0.096129
Feb         0.067143
Mar         0.061613}


### concat multi-index-row using dict

In [96]:
rain1314_by_row_DICT=pd.concat(rain_dict)
rain1314_by_row_DICT

Precipitation
     Month               
2013 Jan         0.096129
     Feb         0.067143
     Mar         0.061613
2014 Jan         0.096129
     Feb         0.067143
     Mar         0.061613

### concat multi-index-col using dict

In [98]:
rain1314_by_col_DICT=pd.concat(rain_dict, axis=1)
rain1314_by_col_DICT

2013          2014
      Precipitation Precipitation
Month                            
Jan        0.096129      0.096129
Feb        0.067143      0.067143
Mar        0.061613      0.061613

Neat ha!

## 6.2.4. Outer & inner joins

### Stacking arrays
    
Let's create some arrays to play with:

In [100]:
import numpy as np
import pandas as pd

In [101]:
A = np.arange(8).reshape(2,4) + 0.1
print(A)

[[0.1 1.1 2.1 3.1]
 [4.1 5.1 6.1 7.1]]


In [102]:
B = np.arange(6).reshape(2,3) + 0.2
print(B)

[[0.2 1.2 2.2]
 [3.2 4.2 5.2]]


In [103]:
C = np.arange(12).reshape(3,4) + 0.3
print(C)

[[ 0.3  1.3  2.3  3.3]
 [ 4.3  5.3  6.3  7.3]
 [ 8.3  9.3 10.3 11.3]]


### Horizontal stacking

Is possible only when the number of rows of each array is the same size i.e. 

We can stack A and B horizontally but **NOT VERTICALLY**

We can stack A and C vertically but **NOT HORIZONTALLY**

We can NOT STACK B and C  **(NOT VERTICALLY , NOT HORIZONTALLY)**

syntax:

    np.hstack([B,A])
    
    or 
    
    np.concatenate([B,A], axis=1)

### Stacking arrays horizontally

In [104]:
np.hstack([B,A])

array([[0.2, 1.2, 2.2, 0.1, 1.1, 2.1, 3.1],
       [3.2, 4.2, 5.2, 4.1, 5.1, 6.1, 7.1]])

In [105]:
# also
np.concatenate([B,A], axis=1)

array([[0.2, 1.2, 2.2, 0.1, 1.1, 2.1, 3.1],
       [3.2, 4.2, 5.2, 4.1, 5.1, 6.1, 7.1]])

### Stacking arrays vertically

In [107]:
np.vstack([A,C])

array([[ 0.1,  1.1,  2.1,  3.1],
       [ 4.1,  5.1,  6.1,  7.1],
       [ 0.3,  1.3,  2.3,  3.3],
       [ 4.3,  5.3,  6.3,  7.3],
       [ 8.3,  9.3, 10.3, 11.3]])

In [108]:
# also
np.concatenate([A,C], axis=0)

array([[ 0.1,  1.1,  2.1,  3.1],
       [ 4.1,  5.1,  6.1,  7.1],
       [ 0.3,  1.3,  2.3,  3.3],
       [ 4.3,  5.3,  6.3,  7.3],
       [ 8.3,  9.3, 10.3, 11.3]])

### Stacking with Incompatible array dimensions

In [111]:
#np.hstack([B,C])

#returns ValueError 

In [112]:
#np.vstack([B,A])
#returns ValueError 

### Population & unemployment data

In [117]:
population = pd.read_csv('pop.csv',index_col=0)
population

2010  Sensus Population
Zip Code ZCTA                         
57538                              322
59916                              130
37660                            40038
2860                             45199

In [116]:
unemployment = pd.read_csv('unemployment.csv',index_col=0)
unemployment

unemployment  participants
Zip                              
2860           0.11         34447
46167          0.02          4800
1097           0.33            42
80808          0.07          4310

### Converting DataFrames to arrays

If we convert DataFrames to arrays we will loose all the labels

In [118]:
population_array = np.array(population) # index info is lost
print(population_array)

[[  322]
 [  130]
 [40038]
 [45199]]


In [120]:
unemployment_array = np.array(unemployment)
print(unemployment_array)

[[1.1000e-01 3.4447e+04]
 [2.0000e-02 4.8000e+03]
 [3.3000e-01 4.2000e+01]
 [7.0000e-02 4.3100e+03]]


### Manipulating data as arrays

Joining DataFrames with different labels with out any sort of label is meaning less!

So if we are joining we must do it based on related index lebels.

In [124]:
#This join is meaningless
print(np.concatenate([population_array, unemployment_array],axis=1)) 

[[3.2200e+02 1.1000e-01 3.4447e+04]
 [1.3000e+02 2.0000e-02 4.8000e+03]
 [4.0038e+04 3.3000e-01 4.2000e+01]
 [4.5199e+04 7.0000e-02 4.3100e+03]]


### Joins: 


### Conactenation and Inner Join: intersection

Intersection of index sets only

It's meaningful join

Intersection of index sets (only common labels)

In [126]:
population

2010  Sensus Population
Zip Code ZCTA                         
57538                              322
59916                              130
37660                            40038
2860                             45199

In [127]:
unemployment

unemployment  participants
Zip                              
2860           0.11         34447
46167          0.02          4800
1097           0.33            42
80808          0.07          4310

In [138]:
pd.concat([population, unemployment], axis=1, join='inner')

2010  Sensus Population  unemployment  participants
2860                    45199          0.11         34447

**If we perform inner join vertically**, it will result in an empty dataFrame, this is because the column lebels are not the same. If there are common column lebels, the vertical inner join will give as values for columns that have common col lebels!

all values will be lost. It gives us only index lebels, which is alone useless!

In [137]:
pd.concat([population, unemployment], join='inner', axis=0)

Empty DataFrame
Columns: []
Index: [57538, 59916, 37660, 2860, 2860, 46167, 1097, 80808]

As we can see here, inner join horizontally finds an index lebel that is present on both population and unemployment and will do the concatenation.

### Conactenation and outer Join: union

`join='outer'` is a default parameter

If the index lebel is present in all dataFrames the merging will be seamless.

A NaN value will be given to a DataFrame which didn't have common index lebels.

e.g an index lebel `2860` is present on both DataFrames `population` and `unemployment` and therefore outer join will fill the entire row with values, but when an index lebel is missing a corrosponding value will be NaN.

In [131]:
pd.concat([population, unemployment], axis=1, join='outer')

2010  Sensus Population  unemployment  participants
1097                       NaN          0.33          42.0
2860                   45199.0          0.11       34447.0
37660                  40038.0           NaN           NaN
46167                      NaN          0.02        4800.0
57538                    322.0           NaN           NaN
59916                    130.0           NaN           NaN
80808                      NaN          0.07        4310.0